include/cppkafka/utils/buffered_producer.h

```diff
@@ -329,26 +418,52 @@ template <typename BufferType>
 BufferedProducer<BufferType>::BufferedProducer(Configuration config)
 : producer_(prepare_configuration(std::move(config))) {
     producer_.set_payload_policy(get_default_payload_policy<BufferType>());
+#ifdef KAFKA_TEST_INSTANCE
+    test_params_ = nullptr;
+#endif
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::add_message(const MessageBuilder& builder) {
+    add_tracker(const_cast<MessageBuilder&>(builder));
     do_add_message(builder, MessagePriority::Low, true);
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::add_message(Builder builder) {
+    add_tracker(builder);
     do_add_message(move(builder), MessagePriority::Low, true);
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::produce(const MessageBuilder& builder) {
-    produce_message(builder);
+    add_tracker(const_cast<MessageBuilder&>(builder));
+    async_produce(builder, true);
+}
+
+template <typename BufferType>
+void BufferedProducer<BufferType>::sync_produce(const MessageBuilder& builder) {
+    TrackerPtr tracker = add_tracker(const_cast<MessageBuilder&>(builder));
```

Comment:
I think there's a general issue now with these member functions taking const refs and using `const_cast` here. What happens if you want to produce the same `MessageBuilder` twice? The second time, `add_tracker` will return `nullptr` so you won't be using it. I'm not sure if this would break something but I think all functions that call `add_tracker` should probably remove the tracker before returning (using some RAII wrapper probably). That way the `MessageBuilder`s will be clean afterwards.

The way this works, you can't really call any produce/add_message overloads using the same builder in 2 threads at the same time because they'll be fighting to set the tracker, which goes against what I'd think given the functions take const refs.

As I see it, `MessageBuilder::internal` is only used in `BufferedProducer` (also via `MessageInternalGuard` but that one's only used in that class as well). Can't you just wrap these builders in a thin template struct that contains the original builder (probably allowing to pass along reference types so you can avoid copies) along with the tracker which will be used internally? That way you won't modify the builders at all and you'll keep the same behavior.

Reply:
This is a tricky scenario with a single MessageBuilder produced multiple times. You're absolutely right that right now two threads will compete for the internal structure, but consider this: imagine you build a MessageBuilder from a Message object. The Message::internal gets transposed inside the MessageBuilder::internal, so now if you want to produce the same MessageBuilder instance multiple times you will be using the same Internal data whatever that is (Tracker in this case or other stuff). Then this is no different from producing multiple MessageBuilders (with initial empty Internal data) and eventually the first thread gets to set the Internal data which will be used again and again.  So what should be the real behavior here? Also if you produce the same Message multiple times, the behavior is identical, you only use the _same_Internal data instance each time. Maybe the behavior should be:
* If I have a MessgeBuilder with empty internal data, then `add_tracker` should behave like RAII. 
* If MessageBuilder already has some internal data then `add_tracker` should be a no-op.